✨Add initial span interface

mayurkale22 opened the pull request:
Completes #29 

This is based on the specs: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#span

This PR contains following interfaces:
- Span
- SpanContext
- Status
- Link
- Attributes
- TraceOptions
- TraceState
- SpanKind

mayurkale22 commented:
@rochdev @vmarchaud @hekike @bg451 Please review. Also, please file an issue [here](https://github.com/open-telemetry/community/issues/new?assignees=&labels=area%2Fgithub-membership&template=membership.md&title=REQUEST%3A+New+membership+for+%3Cyour-GH-handle%3E) so that I can add you as a reviewer.

hekike commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,105 @@
+   */
+  addEvent(
+    name: string,
+    attributes?: { [key: string]: string | number | boolean }
```
Is Object not supported here intentionally?
spec: `values may be string, booleans or numeric type.`
maybe would be a useful addition like we did for OpenCensus's `setAttribute ` and automatically `JSON.stringify()` objects.
I assume `setAttribute` will do the same as spec is `must be either a string, a boolean value, or a numeric type.`

mayurkale22 replied:
Added the `object` type.

hekike commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,105 @@
+   */
+  addLink(
+    spanContext: SpanContext,
+    attributes?: { [key: string]: string | number | boolean }
```
Object?

vmarchaud commented:
@mayurkale22 Which sponsor should we put on the issue ? I guess the first one is you but for the second one ?

danielkhan commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   * @param value the value for this attribute. If the value is a typeof object
+   *     it has to be JSON.stringify-able otherwise it will raise an exception.
+   */
+  setAttribute(key: string, value: string | number | boolean | object): void;
```
If those setters would return the `Span` (`this`), we could support fluid interfaces, like `span.setAttibute().addEvent()`.

rochdev replied:
I think `value` should be of type `any` similar to OpenTracing. The idea is that vendors can accept different values in different ways, so this should not be restricted.

rochdev replied:
There should also be `setAttributes` and `setEvents` to make it easier to add in bulk.

mayurkale22 replied:
> If those setters would return the Span itself (`this`), we would provide fluid interfaces, like `span.setAttibute().addEvent()`.

Sure, done.

mayurkale22 replied:
> There should also be `setAttributes` and `setEvents` to make it easier to add in bulk.

Feel free to open an issue to discuss and if goes well, will include in the next iteration of the API.

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   * Sets an attribute to the span.
+   *
+   * @param key the key for this attribute.
+   * @param value the value for this attribute. If the value is a typeof object
```
This would be problematic for Datadog since we currently accept values that are not JSON serializable and we have a custom serializer that knows how to handle them.

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   */
+  addEvent(
+    name: string,
+    attributes?: { [key: string]: string | number | boolean | object }
```
Similar to the above, `attributes` should be of type `any`.

mayurkale22 replied:
I am fine with changing it to `any` type, but will lose type-safety and to mitigate this I have to add TSLint rule flag (`tslint:disable-next-line:no-any`). WDYT?

rochdev replied:
Since there is a limited number of types in JS it would be possible to simply specify all of them, but I think `any` is clearer that the intent is to accept anything. I'm personally fine with a TSLint comment. If it happens to much we could simply remove the rule. Code reviews will ensure that there is no abuse.

mayurkale22 replied:
Ok, done.

draffensperger replied:
Could we use `unknown` instead of `any`? That allows the same semantics (the passed in value can be any type), but it requires type guards or a cast to actually do something with the type and so is safer from a type checking perspective. As a plus it doesn't require the lint comment. (Most times where you say `any` you can usually say `unknown` and get the same effect)

rochdev replied:
Would it make more sense then to include every possible JavaScript types? There are not that many and it would remove the need for casting for common types.

mayurkale22 replied:
> Could we use unknown instead of any?

For now, I have changed to `unknown `. If needed, will change to either `any` or every possible JavaScript types.

draffensperger replied:
My understanding is that `unknown` is basically equivalent (maybe exactly equivalent?) to the union type of every possible type. You can use `typeof` and `instanceof` that would do runtime checks but are also used by the TS compiler to infer a more narrow type at compile time for code in code blocks guarded by those checks. Casts would only be needed if you somehow know what type it is and force it to that, which perhaps something vendor-specific might do.

See the `function f20(x: unknown)` example in the  in the [TS 3.0 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html) for an example of how `typeof` and `instanceof` can be used for type narrowing by the compiler (and of course checking of the value's actual JS runtime type so it can be correctly serialized).

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   * Adds a link to the Span.
+   *
+   * @param spanContext the context of the linked span.
+   * @param [attributes] the attributes that will be added; these are
```
It's not clear what these attributes are used for.

mayurkale22 replied:
I think `attributes ` are the collection of key:value pairs associated with the link to conserve meaningful information related to link.

rochdev replied:
What I meant is that this should be better explained in the TypeDoc. In general I feel the documentation of the file doesn't explain much of what the different methods do. Of course, documentation can be improved in later iterations.

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   *     associated with this link.
+   */
+  addLink(
+    spanContext: SpanContext,
```
This should accept a `Span` as well.

rochdev replied:
How is the type of link determined?

mayurkale22 replied:
> How is the type of link determined?

Good Point! AFAIK will use `attributes` to highlight the link type and will handle in `sdk` implementation. @songy23 and @rghetia to add more on this.

> This should accept a Span as well.

Can you please mention the use case with `span` ?

rochdev replied:
> Can you please mention the use case with span ?

In general, I think everywhere a `SpanContext` is accepted should also accept a `Span`. This avoids the caller having to always call `span.context()` if they are referencing a span. I don't have a specific use case but I've had a lot of cases where I was happy that OpenTracing did that for me. This is especially true if the context propagator returns the active span.

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   *
+   * @param name the Span name.
+   */
+  updateName(name: string): void;
```
I would use `setName` to be consistent with `setAttribute`.

mayurkale22 replied:
SGTM https://github.com/open-telemetry/opentelemetry-specification/issues/119.

mayurkale22 replied:
The decision is to go with `updateName` for this feature, so terminology will be aligned across languages. I will add TODO with issue number to consider this later.

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   * Call to End of a Span MUST not have any effects on child spans. Those may
+   * still be running and can be ended later.
+   */
+  end(): void;
```
This should accept a timestamp since it's the only way to support some frameworks (i.e. Apollo Tracing).

mayurkale22 replied:
Makes sense, done.

hekike replied:
Shouldn't we change the spec for this?
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#end
> There MUST be no parameter.

rochdev replied:
I would say yes as the same issue applies to every language. The same will be true also for the ability to set the start time.

mayurkale22 replied:
https://github.com/open-telemetry/opentelemetry-specification/pull/120

rochdev commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,106 @@
+   * @returns true if this Span is active and recording information like events
+   * with the AddEvent operation and attributes using setAttributes.
+   */
+  isRecordingEvents(): boolean;
```
From the description it sounds like the fact that it's recording events is only part of the state. Should this be called `isActive` instead or something generic?

mayurkale22 replied:
This flag is used to indicate whether events should be recorded or not. If it's set, `Events` and `Attributes`, etc. will be recorded. Otherwise, these events will be dropped. The only way to set this flag when we create the span. If not set, the implementation will provide a default.

rochdev commented on packages/opentelemetry-types/src/trace/span_context.ts:
```diff
@@ -0,0 +1,30 @@
+  /** The ID of the Span. */
+  spanId: string;
+  /** Trace options to propagate. */
+  traceOptions?: number;
```
It's unclear how this is supposed to be used. What does this number represent?

mayurkale22 replied:
Addressed in 52e53abd9d8e93f470bcba66db47d3edf39fe242, PTAL

rochdev replied:
Would it make more sense to have different fields instead of a single bit field? While it requires a little bit more memory, it makes such an interface a lot easier to understand and use.

rochdev commented on packages/opentelemetry-types/src/trace/span_context.ts:
```diff
@@ -0,0 +1,30 @@
+  /** Trace options to propagate. */
+  traceOptions?: number;
+  /** Tracing-system-specific info to propagate. */
+  traceState?: string;
```
It's unclear what should be the format of this string.

mayurkale22 replied:
Addressed in 52e53abd9d8e93f470bcba66db47d3edf39fe242, PTAL

rochdev replied:
This would be easier to interact with by using an actual object than can then be serialized to a string by some formatter/encoder. Of course this could be done with a getter and a corresponding `_traceState` private property to hold the object, but then every vendor will have to replicate this logic.

mayurkale22 replied:
Makes sense, added `TraceState` and `TraceOptions` classes with TODO comment.

rochdev commented on packages/opentelemetry-types/src/trace/status.ts:
```diff
@@ -0,0 +1,161 @@
+}
+
+/** An enumeration of canonical status codes. */
+export enum CanonicalCode {
```
Does it make sense to have a fixed enum for this? What if the reason falls outside of the range of values?

Also, should there be a code for exceptions, with the error information attached?

mayurkale22 replied:
I think will revise after https://github.com/open-telemetry/opentelemetry-specification/issues/59. Currently it is copied from OpenCensus project.

draffensperger commented on packages/opentelemetry-types/src/trace/span.ts:
```diff
@@ -0,0 +1,109 @@
+   * Do not return `this`. The Span generally should not be used after it
+   * is ended so chaining is not desired in this context.
+   *
+   * @param [endTime] the timestamp to set as Span's end time. If not provided,
```
Could you specify how the timestamp is formatted? Are we going to use epoch millis?

mayurkale22 replied:
If you don't mind, I will update once we have some conclusion here https://github.com/open-telemetry/opentelemetry-node/issues/19

draffensperger replied:
Sounds good. Could you add a TODO comment that points to that issue though?

draffensperger commented on packages/opentelemetry-types/src/trace/span_context.ts:
```diff
@@ -0,0 +1,51 @@
+ * propagated along side of a distributed context.
+ */
+export interface SpanContext {
+  /** The ID of the trace that this span belongs to. */
```
Should this specify the format? Hex-encoded 128-bit number? Same for spanId.

mayurkale22 replied:
Done 424c9fde98a16f69f6893f6e5550695be2c9cb32

mayurkale22 commented:
@hekike @rochdev @danielkhan @draffensperger I tried to address all the comments, please take a look again. I think this is just first iteration will update/add moving forward.

rochdev commented on packages/opentelemetry-types/src/trace/link.ts:
```diff
@@ -0,0 +1,29 @@
+  /** The SpanContext of a linked span. */
+  spanContext: SpanContext;
+  /** A set of attributes on the link. */
+  attributes: Attributes;
```
Should this be optional?

draffensperger commented on packages/opentelemetry-types/src/trace/span_context.ts:
```diff
@@ -0,0 +1,63 @@
+   * sampled or not.
+   * SAMPLED_VALUE = 0x1 and NOT_SAMPLED_VALUE = 0x0;
+   */
+  traceOptions?: TraceOptions;
```
Could this just be a `number` instead of a `TraceOptions` structure? Seems like the comment implies that.

draffensperger commented on packages/opentelemetry-types/src/trace/span_context.ts:
```diff
@@ -0,0 +1,63 @@
+   *     Multiple tracing systems (with different formatting):
+   *         tracestate: rojo=00f067aa0ba902b7,congo=t61rcWkgMzE
+   */
+  traceState?: TraceState;
```
Seems that the comment about is describing this as a string, but the type seems to imply something more complex - is this meant to become the parsed trace state?

I'm OK with leaving this as-is though since the structure is marked as TODO anyway.

draffensperger commented on packages/opentelemetry-types/src/trace/traceoptions.ts:
```diff
@@ -0,0 +1,19 @@
+ * limitations under the License.
+ */
+
+export class TraceOptions {
```
Should the file name be something like `trace_options.ts` to match this?

Also, should this be an interface instead?

draffensperger commented on packages/opentelemetry-types/src/trace/tracestate.ts:
```diff
@@ -0,0 +1,19 @@
+ * limitations under the License.
+ */
+
+export class TraceState {
```
Similar comments here: should the file be `trace_state.ts`, and should this be an interface? (I realize this is TODO, but wanted to comment anyway on this since this is the `types` package)

mayurkale22 commented:
I think I addressed all of the comments by updating the PR, adding TODOs, or opening issues. Merging this PR now. Thanks for the reviews @danielkhan @rochdev @hekike @draffensperger
